federation/schema.go

```diff
@@ -0,0 +1,161 @@
+package federation
+
+import (
+	"bytes"
+	"crypto/sha256"
+	_ "embed"
+	"errors"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	_ "github.com/ipld/go-ipld-prime/codec/dagcbor"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/record"
+	"github.com/multiformats/go-multicodec"
+)
+
+const (
+	stateHeadRecordDomain = "indexer"
+	stateHeadRecordCodec  = "/indexer/fed/stateHeadSignature"
+)
+
+var (
+	Prototypes struct {
+		Head                     schema.TypedPrototype
+		Snapshot                 schema.TypedPrototype
+		ProvidersIngestStatusMap schema.TypedPrototype
+		IngestStatus             schema.TypedPrototype
+
+		link ipld.LinkPrototype
+	}
+
+	//go:embed schema.ipldsch
```

Comment:
This is an interesting pattern - does it impact startup time to do this on startup rather than a compile time step?

Reply:
I have not measured it. I also have not heard complains about its current usage elsewhere being slow.

Happy to measure it if you think it's a blocker.